util/src/hashdb.rs

```diff
@@ -15,13 +15,70 @@
 // along with Parity.  If not, see <http://www.gnu.org/licenses/>.
 
 //! Database of byte-slices keyed to their Keccak hash.
+
+use elastic_array::ElasticArray128;
 use hash::*;
 use std::collections::HashMap;
-use elastic_array::ElasticArray128;
 
 /// `HashDB` value type.
 pub type DBValue = ElasticArray128<u8>;
 
+/// Supplemental trait to `HashDB` supplying generic methods for trait objects, internally
+/// calling monomorphic forms of these methods.
+pub trait HashDBExt {
+	/// Get a reference, transform it, and return the result (prevents cloning).
+	fn get_with<Out, F: FnOnce(&[u8]) -> Out>(&self, _: &H256, _: F) -> Option<Out>;
+}
+
+macro_rules! get_with_fn_def {
+	() => {
+		fn get_with<Out, F: for<'a> FnOnce(&'a [u8]) -> Out>(
+			&self,
+			key: &H256,
+			f: F,
+		) -> Option<Out> {
+			let mut o_func: Option<F>   = Some(f);
+			let mut output: Option<Out> = None;
+
+			{
+				let mut wrapper = |key: &[u8]| {
+					// Unwrap here because `get_exec` is required to call the closure no more than
+					// once.
+					output = Some((o_func.take().unwrap())(key));
```

Comment:
could you move the comment into an `expect`? 
`Thread N panicked at "Called Option::unwrap on a None value", Option.rs:999`
is way worse than
`Thread N panicked at "this exact, expected invariant was violated!", Option.rs:999`.

We don't always get backtraces from our issue reports.